src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java

```diff
@@ -0,0 +1,183 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.snowflake.kafka.connector.internal.streaming.telemetry;
+
+import static com.snowflake.kafka.connector.internal.metrics.MetricsUtil.constructMetricName;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.snowflake.kafka.connector.internal.KCLogger;
+import com.snowflake.kafka.connector.internal.metrics.MetricsJmxReporter;
+import com.snowflake.kafka.connector.internal.metrics.MetricsUtil;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class SnowflakeTelemetryChannelStatus {
+  private static final KCLogger LOGGER =
+      new KCLogger(SnowflakeTelemetryChannelStatus.class.toString());
+
+  public static final long NUM_METRICS = 3; // update when new metrics are added
+  private static final long INITIAL_OFFSET = -1;
+
+  // channel properties
+  private final String channelName;
+  private final MetricsJmxReporter metricsJmxReporter;
+
+  // offsets
+  private AtomicLong offsetPersistedInSnowflake;
+  private AtomicLong processedOffset;
+  private AtomicLong latestConsumerOffset;
+
+  /**
+   * Creates a new object tracking {@link
+   * com.snowflake.kafka.connector.internal.streaming.TopicPartitionChannel} metrics with JMX
+   * TODO @rcheng: update comment when extends telemetryBasicInfo
+   *
+   * @param tableName the table the channel is ingesting to
+   * @param channelName the name of the TopicPartitionChannel to track
+   * @param enableCustomJMXConfig if JMX metrics are enabled
+   * @param metricsJmxReporter used to report JMX metrics
+   */
+  public SnowflakeTelemetryChannelStatus(
+      final String tableName,
+      final String channelName,
+      final boolean enableCustomJMXConfig,
+      final MetricsJmxReporter metricsJmxReporter) {
+    this.channelName = channelName;
+    this.metricsJmxReporter = metricsJmxReporter;
+
+    this.offsetPersistedInSnowflake = new AtomicLong(INITIAL_OFFSET);
+    this.processedOffset = new AtomicLong(INITIAL_OFFSET);
+    this.latestConsumerOffset = new AtomicLong(INITIAL_OFFSET);
+    if (enableCustomJMXConfig) {
+      if (metricsJmxReporter == null) {
+        LOGGER.error("Invalid metrics JMX reporter, no metrics will be reported");
+      } else {
+        this.registerChannelJMXMetrics();
+      }
+    }
+  }
+
+  /** Registers all the Metrics inside the metricRegistry. */
+  private void registerChannelJMXMetrics() {
+    LOGGER.debug(
+        "Registering new metrics for channel:{}, removing existing metrics:{}",
+        this.channelName,
+        this.metricsJmxReporter.getMetricRegistry().getMetrics().keySet().toString());
+    this.metricsJmxReporter.removeMetricsFromRegistry(this.channelName);
+
+    MetricRegistry currentMetricRegistry = this.metricsJmxReporter.getMetricRegistry();
+
+    try {
+      // offsets
+      currentMetricRegistry.register(
+          constructMetricName(
+              this.channelName,
```

Comment:
Not familiar with JMX, but can a channel name be uniquely identify a metric? What if we have two channels with the same name but different client sequencer?

Reply:
Channel names are meant to be[ unique identifiers](https://github.com/snowflakedb/snowflake-kafka-connector/blob/3879f8a3c2349b1ab3955c0bb22e601e08ba35da/src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java#L507) for that particular topic and partition. 

From what I understand, client sequencers are bumped up when a channel is reopened, which means the registry should replace the previous channel name's metrics. This should be ok behavior, wdyt?

Another potential issue is two connectors listening to the same topic, however this is not an issue because the MetricRegistry is created off the unique connector name.

Reply:
what about during rebalance? Any concern about emit metrics from two nodes?

Reply:
There shouldn't be rebalance issues because JmxReporters are created at the serviceSink level. When registering new metrics for a TopicPartitionChannel during rebalance, we remove all existing metrics and then register new ones.